browser-extension/mv2/src/background/background.js

```diff
@@ -1,3 +1,75 @@
+// import { startEventBatchWriter } from "./EventWriter";
+
+/* can be moved to a different file --- EventWriter.js */
+
+let eventsBatch = [];
+let batchWriterInterval = null;
+
+const LOCAL_ANALYTICS_KEY = "analytics_event";
+
+function addEventToBatch(event) {
+  // export
+  // todo: generate uuid for each event
+  eventsBatch.push(event);
+}
+
+async function writeBatchToLocalStorage() {
+  const batchedEvents = [...getBatchedEvents()];
+
+  if (batchedEvents.length) {
+    // write batched events to local storage
+    await RQ.StorageService.getRecord(LOCAL_ANALYTICS_KEY).then(
+      async (storageEvents) => {
+        if (storageEvents) batchedEvents.push(...storageEvents);
+
+        const newAnalyticsEntry = {};
+        newAnalyticsEntry[LOCAL_ANALYTICS_KEY] = batchedEvents;
+        await RQ.StorageService.saveRecord(newAnalyticsEntry).then((_) =>
+          console.log("BG to LOCAL: Events Write complete")
+        );
+      }
+    );
+    // now these events are considered ready to be sent to UI
+    // todo: call sendEvents() // being implemented separately by @nafees
+  }
+}
+
+function getBatchedEvents() {
+  // remove from the local batched array
+  const batchedEvents = [...eventsBatch];
+  eventsBatch = [];
+  return batchedEvents;
+}
+
+function startEventBatchWriter(intervalTime = 2000) {
+  // export
+  if (!batchWriterInterval) {
+    batchWriterInterval = setInterval(async () => {
+      await writeBatchToLocalStorage();
+    }, intervalTime);
+  }
+
+  return batchWriterInterval;
+}
+
+async function getEvents() {
+  // export
+  const events = RQ.StorageService.getRecord(LOCAL_ANALYTICS_KEY)
```

Comment:
```suggestion
  const events = 
await RQ.StorageService.getRecord(LOCAL_ANALYTICS_KEY)
```
or should use try-catch

Reply:
can't see any difference either way, we just need a way to return null or empty array when events are not stored.
Both work the same IMO so not changing for now. 
as for awaiting, this returns the promise to the main action of the function so for now, again doesn't make a difference